File: zadapter3/src/main/java/com/zone/adapter3/absorb/StickyOnScrollListener.java
package com.zone.adapter3.absorb;

import android.graphics.Color;
import android.support.annotation.ColorInt;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import com.zone.adapter3.QuickConfig;
import com.zone.adapter3.bean.StickyViewDelegates;

/**
 * [2018] by Zone
 * <p>
 */
public class StickyOnScrollListener extends RecyclerView.OnScrollListener {

    private final FrameLayout vpShow;
    private StickyViewDelegates[] mStickyViews;
    private int[] stickyPostions;
    private View placeholderView;
    private int preStickyIndex = -1;


    public StickyOnScrollListener(FrameLayout vpShow, int[] stickyPostions, StickyViewDelegates[] mStickyViews) {
        this.vpShow = vpShow;
        this.stickyPostions = stickyPostions;
        this.mStickyViews = mStickyViews;
    }

    @Override
    public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
        super.onScrollStateChanged(recyclerView, newState);
    }

    int placeColor =Color.TRANSPARENT;
    public void setPlaceColor(@ColorInt int color){
        this.placeColor =color;
    }

    @Override
    public synchronized void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);

        int pos = 0;
        try {
            //GridLayoutManager 继承LinearLayoutManager 所以也支持GridLayoutManager
            pos = ((LinearLayoutManager) (recyclerView.getLayoutManager())).findFirstVisibleItemPosition();
            QuickConfig.e("吸附-》findFirstVisibleItemPosition：" + pos);
        } catch (Exception e) {
            e.printStackTrace();
            throw new IllegalStateException("仅仅支持继承LinearLayoutManager与GridLayoutManager的布局!");
        }

        int showStickyPos = findShowPos(pos);
        //找到对应的 数组位置  用他去找view

        int nowStickyIndex = 0;
        for (int i = 0; i < stickyPostions.length; i++) {
            if (showStickyPos == stickyPostions[i]) {
                nowStickyIndex = i;
                break;
            }
        }

        if (showStickyPos == -1) {//啥也不显示
            if (preStickyIndex != -1) {
                restorePreSticky();
                QuickConfig.e("吸附-》未找到吸附位置：重置占位头");
                preStickyIndex = -1;
            }
        } else {
            if (placeholderView == null) {
                placeholderView = new View(recyclerView.getContext());
                placeholderView.setBackgroundColor(placeColor);
            }


            if (preStickyIndex == -1 || preStickyIndex != nowStickyIndex) {
                //如果当前吸附的与吸附的判断是一样的话 跳过 因为已经设置过了,或者preStickyIndex=-1代表之前没有设置过

                if (preStickyIndex != -1) {
                    //找到之前那个itemView 还原了
                    restorePreSticky();
                }

                QuickConfig.e("吸附-》吸附位置:" + nowStickyIndex + "狸猫换太子");
                //ItemView的狸猫换太子 显示在VP中
                View stickyView = mStickyViews[nowStickyIndex].getEHFHolder().itemView;
                removeParent(stickyView);
                removeParent(placeholderView);
                placeholderView.setLayoutParams(stickyView.getLayoutParams());
                //设置站位高度
                placeholderView.getLayoutParams().height = stickyView.getHeight();

                if (mStickyViews[nowStickyIndex].getParentHolder() != null) {
                    ViewGroup recyclerHolder = (ViewGroup) mStickyViews[nowStickyIndex].getParentHolder().itemView;
                    recyclerHolder.addView(placeholderView);
                    mStickyViews[nowStickyIndex].setPlaceholderView(placeholderView);
                    vpShow.addView(stickyView);
                    preStickyIndex = nowStickyIndex;

                }
            }

            // 位移计算!
            if (showStickyPos != stickyPostions[stickyPostions.length - 1]) {//最后的那个不进行位移
                int next = stickyPostions[nowStickyIndex + 1];
                View stickyView = mStickyViews[nowStickyIndex].getEHFHolder().itemView;
                if (stickyView != null) {
                    if (recyclerView.findViewHolderForLayoutPosition(next) == null)
                        stickyView.setTranslationY(0);
                    else {
                        View targetView = recyclerView.findViewHolderForLayoutPosition(next).itemView;
                        if (targetView.getTop() <= stickyView.getHeight())
                            stickyView.setTranslationY(targetView.getTop() - stickyView.getHeight());
                    }
                }

            }

        }
    }

    private void restorePreSticky() {
        QuickConfig.e("吸附-》还原了之前那个位置：" + stickyPostions[preStickyIndex]);
        View preStickyView = mStickyViews[preStickyIndex].getEHFHolder().itemView;
        removeParent(preStickyView);

        removeParent(placeholderView);
        mStickyViews[preStickyIndex].setPlaceholderView(null);

        preStickyView.setTranslationY(0);
        ViewGroup preVp = (ViewGroup) mStickyViews[preStickyIndex].getParentHolder().itemView;
        preVp.addView(preStickyView);
    }

    private void removeParent(View view) {
        ViewGroup vp = (ViewGroup) (view.getParent());
        if (vp != null)
            vp.removeView(view);
    }

    // 3,6,9为例  [1,3)show 无 ,[3,6)show 3,[6,9)show 6,>=9 show 9,
    private int findShowPos(int pos) {
        // if (itemViews[i - 1].getTop() <= 0) {//注意:getTop不包括 dector

        for (int i = 0; i < stickyPostions.length; i++) {
            if (pos < stickyPostions[i]) {
                if (i == 0)//[1,3)show 无
                    return -1;
                //[3,6)show 3,[6,9)show 6
//                System.out.println("Absorb_ i:" + i + "\t itemViews[i-1].getTop()" + itemViews[i - 1].getTop());
                return findWithOutDecTop(i);
            }
        }
        // >9 show 9   =9的时候看top ,如果top<=0 就是9 如果top>
        if (pos == stickyPostions[stickyPostions.length - 1]) {
            return findWithOutDecTop(stickyPostions.length);
        } else
            return stickyPostions[stickyPostions.length - 1];

    }

    // 主要是处理 dec的  如果3是断头，但是3有dec的话，getTop<=0才代表断头是3，不然就是3之前的断头
    private int findWithOutDecTop(int i) {
        if (mStickyViews[i - 1].getParentHolder() != null) {
            View itemView = mStickyViews[i - 1].getParentHolder().itemView;
            // getTop如果<=0的时候 才显示这个位置的断头
            if (itemView.getTop() <= 0) {//注意:getTop不包括 dector
                return stickyPostions[i - 1];
            } else {//不然还是显示上一个位置的断头
                if (i - 2 < 0) {//上一个位置是无的情况
                    return -1;
                }
                return stickyPostions[i - 2];
            }
        }
        return -1;
    }

}

File: zadapter3/src/main/java/com/zone/adapter3kt/sticky/StickyOnScrollListener.kt
package com.zone.adapter3kt.sticky

import android.graphics.Color
import android.support.annotation.ColorInt
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.zone.adapter3kt.QuickConfig
import com.zone.adapter3kt.adapter.StickyAdapter

/**
 * [2018] by Zone
 */
class StickyOnScrollListener<T>(private val vpShow: FrameLayout, val adapter: StickyAdapter<T>, @ColorInt val placeColor: Int = Color.TRANSPARENT) : RecyclerView.OnScrollListener() {
    internal var preShowStickyPosi = -1
    private var placeholderView: View? = null
    internal var mNowStickyViewHolder: StickyChildHolder? = null

    override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
        super.onScrollStateChanged(recyclerView, newState)
        if (recyclerView == null) return
    }

    @Synchronized
    override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)
        if (recyclerView == null) return
        //为了 变换的时候不影响这次计算 
        val stickyList = adapter.stickyList
        var pos = 0
        try {
            //GridLayoutManager 继承LinearLayoutManager 所以也支持GridLayoutManager
            pos = (recyclerView.layoutManager as LinearLayoutManager).findFirstVisibleItemPosition()
            QuickConfig.e("吸附-> findFirstVisibleItemPosition：$pos")
        } catch (e: Exception) {
            e.printStackTrace()
            throw IllegalStateException("仅仅支持继承LinearLayoutManager与GridLayoutManager的布局!")
        }

        //找到应该显示的pos
        val (shouldShowStickyPos, nextStickyIndex) = findShowPos(pos, recyclerView, stickyList)
        if (shouldShowStickyPos == -1) { //没找到应该显示的话
            if (preShowStickyPosi != -1) { //之前还有头部显示的话   不显示！
                QuickConfig.e("释放掉 外层显示的view")
                releaseNowShowSticky(recyclerView)
            }
        } else {
            if (placeholderView == null) {
                placeholderView = View(recyclerView.context)
                placeholderView!!.setBackgroundColor(placeColor)
            }

            //如果之前外层显示的与当前显示的pos不一样就更换
            if (preShowStickyPosi != shouldShowStickyPos) {

                //如果之前外层没有显示 就是-1的话就不用释放掉了  因为-1的时候已经释放过了
                if (preShowStickyPosi != -1 && mNowStickyViewHolder != null) releaseNowShowSticky(recyclerView)

                //ItemView的狸猫换太子 显示在VP中
                val shouldViewHolder = recyclerView.findViewHolderForLayoutPosition(shouldShowStickyPos)
                if (shouldViewHolder == null) { //没有  自己创建 ,而不用空白占位
                    val viewType = adapter.getItemViewType(shouldShowStickyPos)
                    val stickyChildHolderItemView: View?
                    for (entry in adapter.stickyViewTypeMap.entries) {
                        if (entry.value == viewType) {
                            val delegate = adapter.delegatesManager.getDelegateNoMap(entry.key) //原始viewtype生成对应的delegate
                            if (delegate == null) throw IllegalStateException(" No delegate found for item ")
                            val innerHolder=adapter.delegatesManager.onCreateViewHolder(vpShow, entry.value)
                            val viewHolder = StickyChildHolder(innerHolder.itemView, entry.key)
//                            val viewHolder = StickyChildHolder(delegate.onCreateViewHolder(vpShow).itemView, entry.key)
                            stickyChildHolderItemView = viewHolder.itemView
                            vpShow.addView(stickyChildHolderItemView)
                            placeholderView!!.layoutParams = stickyChildHolderItemView.layoutParams
                            mNowStickyViewHolder = viewHolder
                            //viewHolder 是StickyChildHolder  而不是StickHolder所以不会被移除parent 仅仅是进行绑定
                            adapter.onBindViewHolder(viewHolder, shouldShowStickyPos, null)
                            break
                        }
                    }
                } else { //有的话 用空白  替换
                    if (shouldViewHolder is StickyHolder) {
                        //安全使用 空白站位View
                        removeParent(placeholderView!!)
                        placeholderView!!.layoutParams = shouldViewHolder.stickyChildHolderItemView.layoutParams
                        placeholderView!!.layoutParams.height = shouldViewHolder.stickyChildHolderItemView.height
                        (shouldViewHolder.itemView as FrameLayout).addView(placeholderView!!)

                        //外层显示 从里层拿出来的viewholder
                        removeParent(shouldViewHolder.stickyChildHolderItemView)
                        vpShow.addView(shouldViewHolder.stickyChildHolderItemView)
                        mNowStickyViewHolder = shouldViewHolder.stickyChildHolder
                    }
                }
                preShowStickyPosi = shouldShowStickyPos
            }

            // 位移计算!
            if (shouldShowStickyPos != stickyList[stickyList.size - 1].posi) {//最后的那个不进行位移
                //下一个的top 和 当前show的高度进行对比
                val nextStickyView = stickyList[nextStickyIndex].posi
                val stickyView = mNowStickyViewHolder!!.itemView
                if (stickyView != null) {
                    if (recyclerView.findViewHolderForLayoutPosition(nextStickyView) == null)
                        stickyView.translationY = 0f
                    else {
                        val targetView = recyclerView.findViewHolderForLayoutPosition(nextStickyView).itemView
                        if (targetView.top <= stickyView.height) stickyView.translationY = (targetView.top - stickyView.height).toFloat()
                        else stickyView.translationY = 0f
                    }
                }
            }
        }
    }

    private fun releaseNowShowSticky(recyclerView: RecyclerView) {
        //release释放的时候 如果在 对应position位置找到的话 ,替换 . 找不到 制空不管即可

        val shouldViewHolder = recyclerView.findViewHolderForLayoutPosition(preShowStickyPosi)
        removeParent(mNowStickyViewHolder!!.itemView)

        if (shouldViewHolder != null && shouldViewHolder is StickyHolder) {
            //找到占位view 移除父亲  甭管他有没有
            removeParent(placeholderView!!)
            //因为有可能 该位置的站位 不是空白  因为如果是scrollto定位的话 往回滚。那么我仅仅在外面生成了view ,但没有用空白去替换啊
            (shouldViewHolder.itemView as ViewGroup).removeAllViews()

            //把sticky设置成当前的
            shouldViewHolder.stickyChildHolder = mNowStickyViewHolder!!
            shouldViewHolder.stickyChildHolderItemView = mNowStickyViewHolder!!.itemView
            //移除外面的view的parent
            removeParent(shouldViewHolder.stickyChildHolderItemView)

            //把外城view 设置到
            shouldViewHolder.stickyChildHolderItemView.translationY = 0F
            (shouldViewHolder.itemView as FrameLayout).addView(shouldViewHolder.stickyChildHolderItemView)
        }
        mNowStickyViewHolder = null
        preShowStickyPosi = -1
    }

    fun childIsPlaceholderView(vp: FrameLayout): Boolean {
        return if (placeholderView != null && placeholderView!!.parent != null && placeholderView!!.parent == vp) true else false
    }

    //当现在在外面显示的posi与adapter bind的posi相同的时候 adapter用占位符占位
    fun adapterAddPlaceHolder(holder:StickyHolder){
        if(mNowStickyViewHolder!=null){
            removeParent(placeholderView!!)
            placeholderView!!.layoutParams = mNowStickyViewHolder!!.itemView.layoutParams
            placeholderView!!.layoutParams.height = mNowStickyViewHolder!!.itemView.height
            (holder.itemView as FrameLayout).addView(placeholderView!!)
        }
    }

    private fun removeParent(view: View) {
        val vp = view.parent
        if (vp != null) (vp as ViewGroup).removeView(view)
    }

    // 3,6,9为例  [0,3)show 无 ,[3,6)show 3,[6,9)show 6,>=9 show 9
    private fun findShowPos(pos: Int, recyclerView: RecyclerView, stickyList: ArrayList<StickyAdapter.StickyEntity<T>>): ShowPos {
        // if (itemViews[i - 1].getTop() <= 0) {//注意:getTop不包括 dector
        for (i in 0 until stickyList.size) {
            if (pos < stickyList[i].posi) {
                return if (i == 0) ShowPos(-1, 0) else ShowPos(stickyList[i - 1].posi, i)
            }
        }
        // 如果posi 大于 列表中最后一个值
        return ShowPos(stickyList[stickyList.size - 1].posi, stickyList.size - 1)
    }

    data class ShowPos(val shouldPos: Int, val nextStickyIndex: Int)
}